### PR TITLE
Move article: first version

### PR DESCRIPTION
**Pull request by the-glu**
This is a proposed feature to move articles in the tree of articles. It's useful eg. after a mediawiki import to organize articles :)

I don't really know if there is no side-issues to this. At least one exist: links are not updated inside others articles of the wiki and there is no redirection left.
I tested to move articles around and I didn't see any incoherence in the database nor partial issues.

I had to include CSS in a `style` tag. This is probably wrong but it's seem to be only a file for bootstrap. Where can I include extra CSS ?

### CODE REVIEW

**Comment by benjaoming** [comment]
Hi @the-glu 

That's ambitious :) But also a feature that is often missed!!

Re: CSS: Everything is written in LESS. You need to do changes [here](https://github.com/django-wiki/django-wiki/blob/master/wiki/static/wiki/bootstrap/less/wiki/wiki.less) and run build-less.sh.

As you say, links will break if there's not a mechanism to fix it properly. Also caches will have to be cleared for article lists to work. In fact, invalidating the cache is non-trivial, and you risk to have to invalidate the whole thing.

Maybe you can respond in #154 as that's the issue where there's some discussion already taking place :)

**Comment by the-glu** [comment]
Okay ! So:

- About the django-mptt/django-mptt#236 I don't think it's should be fixed in django-wiki, right ? Hopefully they will fix the problem, I don't think in a simple django-wiki installation it's an issue (aka moving a article 'alone' won't trigger it).

- I added cache invalidation and it's fix the problem with the article index (and not the TOCs who is only for the article - I assume you made a typo ;) )

- I moved the CSS to the correct part, thanks for the pointer ^^

- About redirection I totally agree. I will try to make a pull request with redirection (maybe today), and then one to add redirection when moving article, but I think this can already be merged when ready, to split issues in different parts :)

**Comment by the-glu** [comment]
Ping ? :)

File: wiki/templates/wiki/move.html [hunk 1]
```diff
@@ -0,0 +1,75 @@
+{% extends "wiki/article.html" %}
+{% load wiki_tags i18n sekizai_tags %}
+
+{% block wiki_pagetitle %}{% trans "Move" %}: {{ article.current_revision.title }}{% endblock %}
+
+{% block wiki_contents_tab %}
+
+{% if not user.is_authenticated %}
```
Comment (benjaoming) [inline wiki/templates/wiki/move.html]: Is this page eligible for anonymous users? Was just wondering if it makes sense or if there's a danger or complexity in allowing that?

Comment (the-glu) [inline wiki/templates/wiki/move.html]: It's just a warning about IP beeing looged - same as article editing.

File: wiki/views/article.py [hunk 2]
```diff
@@ -401,6 +401,80 @@ def get_context_data(self, **kwargs):
...
+        kwargs = self.get_form_kwargs()
+        return form_class(**kwargs)
+
+    def get_context_data(self, **kwargs):
+
+        kwargs['root_path'] = models.URLPath.root()
+
+        return super(Move, self).get_context_data(**kwargs)
+
+    def form_valid(self, form):
```
Comment (benjaoming) [inline wiki/views/article.py]: Maybe this should have an atomic decorator? While the re-parenting in mptt seems fairly atomic, there's a couple of other calls surrounding it.

File: wiki/templates/wiki/move.html [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+</p>
+{% endif %}
+
+<div class="row">
+
+  <div class="col-lg-8">
+    <form method="POST" class="form-horizontal" id="article_move_form">
+      <input type="hidden" name="save" value="1" />
+      <input type="hidden" name="preview" value="1" />
+      {% with form as form %}
```
Comment (benjaoming) [inline wiki/templates/wiki/move.html]: Should this be here? And the two above fields?

File: wiki/templates/wiki/move.html [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+{% block wiki_contents_tab %}
+
+{% if not user.is_authenticated %}
+<p class="alert alert-warning">
+{% trans "<strong>Warning:</strong> You are not logged in. Your IP address will be logged." %}
+<a href="{% url 'wiki:login' %}?next={% url 'wiki:move' article_id=article.id path=urlpath.path %}">{% trans "Click here to log in" %} &raquo;</a>
+</p>
+{% endif %}
+
+<div class="row">
```
Comment (benjaoming) [inline wiki/templates/wiki/move.html]: Maybe a nice little intro about breaking links etc... mentioning the "known issues" about moving articles.

File: wiki/templates/wiki/move.html (hunk 3 above)
Comment (the-glu) [inline wiki/templates/wiki/move.html]: I did exactly as https://github.com/django-wiki/django-wiki/blob/master/wiki/templates/wiki/edit.html#L22 . Should something beeing changed ?

File: wiki/templates/wiki/move.html (hunk 1 above)
Comment (benjaoming) [inline wiki/templates/wiki/move.html]: Yes but the point was more if anonymous users are supposed to see this page?

Comment (the-glu) [inline wiki/templates/wiki/move.html]: I can add a new right, would it be ok ?

**Comment by benjaoming** [comment]
Looks good :) But could you add a test for all this? :)

**Comment by the-glu** [comment]
Ok, I will correct the few remarks and try to add a test :)

File: wiki/templates/wiki/move.html (hunk 3 above)
Comment (benjaoming) [inline wiki/templates/wiki/move.html]: Well, this page is not for editing the article contents, so all of the above is redundant :)

Comment (benjaoming) [inline wiki/templates/wiki/move.html]: ...and below 2 lines :)

File: wiki/forms.py [hunk 5]
```diff
@@ -124,6 +124,11 @@ class CreateRootForm(forms.Form):
         required=False, widget=getEditor().get_widget())  # @UndefinedVariable
 
 
+class MoveForm(forms.Form):
+
+    destination = forms.CharField(label=_('Destination'), initial='', help_text=_('Childrens are also moved. Be carefull, links are NOT updated !'))
```
Comment (benjaoming) [inline wiki/forms.py]: How is it validated that the destination is a valid slug?

File: wiki/views/article.py [hunk 6]
```diff
@@ -401,6 +401,80 @@ def get_context_data(self, **kwargs):
...
+
+        dest_path = get_object_or_404(models.URLPath, pk=form.cleaned_data['destination'])
+
+        if not self.urlpath.parent:
+            messages.error(
+                self.request,
+                _('This article cannot be moved because it is a root article.'))
+            return redirect('wiki:get', article_id=self.article.id)
+
+        dest_is_children = False
```
Comment (benjaoming) [inline wiki/views/article.py]: dest_has_children :)

File: wiki/forms.py (hunk 5 above)
Comment (benjaoming) [inline wiki/forms.py]: I'm a bit confused -- what's this field actually used for? It seems it's for a PK for URLPath? Should it perhaps be made a ModelChoiceField with widget=TextInput() ?

File: wiki/views/article.py [hunk 7]
```diff
@@ -401,6 +401,80 @@ def get_context_data(self, **kwargs):
...
+            ancestor.article.clear_cache()
+
+        self.urlpath.parent = dest_path
+        self.urlpath.save()
+
+        # Reload url path form database
+        self.urlpath = get_object_or_404(models.URLPath, pk=self.urlpath.pk)
+
+        # Use a copy of yourself (to avoid cache) and update articles links
+        # agains
```
Comment (benjaoming) [inline wiki/views/article.py]: Why is necessary to clear the cache one more time? Could the one before be deleted then?

**Comment by ostrokach** [comment]
What is the status of this? 
Being able to move article trees around would be an awesome feature! 
Right now I'm just copy-pasting text...

**Comment by benjaoming** [comment]
@ostrokach - I added the "needs-tests" label. The status is that I'm willing to merge this if there are tests written for special cases like:

 - moving an article to an already existing location
 - ensuring that moving an article does x to already existing links
 - ensuring that moving an article does x to already existing links of all child pages

There are also a couple of unanswered comments.

I don't think this PR is far from being merge-able, but I'm not able to make the effort of finalizing it.

Since I haven't heard from @the-glu for a while, I think it's probably up to others to branch out and make a new PR..

**Comment by benjaoming** [comment]
Closing this PR, partly because we changed away from Coveralls, and party because keeping a very old, unmerged PR seems to imply that django-wiki has a problem accepting PRs :)

But I really hope someone - if not @the-glu - will carry this to the finish line, although I can see that it does need a bit of work to arrive.

Thanks for the work so far!

**Pull request closed without merging**